lib/common.js

```diff
@@ -327,43 +327,77 @@ function matchStringOrRegexp(target, pattern) {
   return pattern instanceof RegExp  ? str.match(pattern) : str === String(pattern);
 }
 
-// return [newKey, newValue]
-function formatQueryValue(key, value, options) {
+function formatQueryParams(queryParams) {
+  var formattedQueryParams = {}
+  for (var q in queryParams) {
+    if (!_.isUndefined(formattedQueryParams[q])) {
+      continue;
+    }
+    formattedQueryParams[q] = formatQueryValue(queryParams[q]);
+  }
+  return formattedQueryParams;
+}
+
+function decodeQueryParams(queryParams) {
+  var decodedQueryParams = {}
+  for (var q in queryParams) {
+    if (!_.isUndefined(decodedQueryParams[q])) {
+      continue;
+    }
+    decodedQueryParams[percentDecode(q)] = formatQueryValue(queryParams[q], percentDecode);
+  }
+  return decodedQueryParams;
+}
+
+function encodeQueryParams(queryParams) {
+  var encodedQueryParams = {}
+  for (var q in queryParams) {
+    if (!_.isUndefined(encodedQueryParams[q])) {
+      continue;
+    }
+    encodedQueryParams[percentEncode(q)] = formatQueryValue(queryParams[q], percentEncode);
+  }
+  return encodedQueryParams;
+}
+
+function formatQueryValue(value, stringFunction) {
   switch (true) {
-    case _.isNumber(value): // fall-though
+    case _.isNumber(value): // fall-through
     case _.isBoolean(value):
       value = value.toString();
       break;
-    case _.isUndefined(value): // fall-though
+    case _.isUndefined(value): // fall-through
     case _.isNull(value):
       value = '';
       break;
     case _.isString(value):
-      if(options.encodedQueryParams) {
-        value = percentDecode(value);
+      if(stringFunction) {
```

Comment:
I think we need a comment here - what happens when there is no `stringFunction` and why? I would maybe rename `stringFunction` to `stringTransformerFn` or something such.